File: run_model_combinations.pl
#!/usr/bin/env perl

# run_model_combinations.pl
# Run a set of combined models using shared_ancestry_simulator.R based on a CSV file of model parameter combinations

# Written for "Evaluating statistics for the identification of introgressed loci"
# by Simon H. Martin, John W. Davey and Chris D. Jiggins
# John Davey:   jd626@cam.ac.uk
# Simon Martin: shm45@cam.ac.uk
# November-December 2013

use strict;
use warnings;
use Carp;
use English;
use Data::Dumper;
use Getopt::Long;


my $numwindows = 1000;
my $windowsize = 5000;
my $modelfile = "";
my $threads = 1;
my $popsize = 1000000;
my $simulator_path = "..";
my $options_okay = GetOptions(
	'numwindows=i' => \$numwindows,
	'windowsize=i' => \$windowsize,
	'modelfile=s' => \$modelfile,
	'threads=i'   => \$threads,
    'popsize=i'   => \$popsize,
    'simulator_path=s' => \$simulator_path,
);

croak "Please supply a CSV file containing model parameters with -m" if $modelfile eq "";

croak "Simulator path $simulator_path does not exist!\n" if !-d $simulator_path;

my $model = <<EOM;
seqgen:
    m: HKY
    l: $windowsize
ms:
    pops:
        - {name: 1, seqs: 8}
        - {name: 2, seqs: 8}
        - {name: 3, seqs: 8}
        - {name: 4, seqs: 8}
    opts:
        genperyear: 4
        n0: $popsize
        mu: 2.5e-9
        split:
            - {time: 3,   pops: [4,1]}
EOM

sub output_models {
    my ($model, $modeltype, $params) = @_;
    my $pops = $modeltype eq "Background" ? "2,1" : $modeltype eq "Alternate" ? "2,3" : "0,0";
    my $printpops = $pops;
    $printpops =~ s/,//g;
    foreach my $t123 (sort {$a<=>$b} keys %{$params}) {
        my $model_t123 = $model . "            - {time: $t123, pops: [3,1]}\n";
        foreach my $txx (keys %{$params->{$t123}}) {
            open my $model_file, ">", "$modeltype\_t123-$t123\_t$printpops-$txx.yml" or croak "Can't open model file: $OS_ERROR!\n";
            my $model_t123_txx = $model_t123 . "            - {time: $txx, pops: [$pops]}\n";
            print $model_file $model_t123_txx;
            close $model_file;
        }
    }
}

open my $combined, "<", $modelfile or croak "Can't open model file: $OS_ERROR!\n";

my $header = <$combined>;
chomp $header;
my @header = split /,/, $header;
my %background;
my %alternate;
my @combinations;
while (my $model = <$combined>) {
    chomp $model;
    my @f = split /,/, $model;
    $background{$f[1]}{$f[2]}++;
    $alternate{$f[3]}{$f[4]}++;
    my %comb;
	$comb{type}=$f[0];
    map {
        my ($type, $var) = split /_/, $header[$_];
        $comb{pars}{$type}{$var} = $f[$_]
    } 1..$#header;
    push @combinations, \%comb;
}

close $combined;

my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime();
$mon++; # originally in 0..11
$year+=1900;
my $outdir = "model_files_win$numwindows\_size$windowsize\_ne$popsize\_$year-$mon-$mday";
mkdir($outdir) if !-d $outdir;
chdir($outdir);

open my $log, ">", "$outdir.log" or croak "Can't open log file! $OS_ERROR\n";

output_models($model, "Background", \%background);
output_models($model, "Alternate", \%alternate);


foreach my $combination (@combinations) {
    my $command = "$simulator_path/shared_ancestry_simulator.R -w $numwindows -T $threads -c ";
    foreach my $model_type (keys %{$combination->{pars}}) {
        my $prop = $model_type eq "Background" ? 0.9 : 0.1;
        $command .= $model_type . "_t123-";
        $command .= $combination->{pars}{$model_type}{t123};
        $command .= "_";
        delete $combination->{pars}{$model_type}{t123};
        my $var2 = (keys %{$combination->{pars}{$model_type}})[0];
        $command .= "$var2-$combination->{pars}{$model_type}{$var2}.yml:$prop,";
    }
    chop $command;
    print $log "$command\t" . localtime() . "\n";
    my $simout = `$command`;
    croak "Couldn't run command: $command\n(Is shared_ancestry_simulator.R in your path?)\n" if !defined $simout;
    croak "Command failed: $command\n$simout\n" if ($simout ne "");
}

File: run_model_combinations.py
#!/usr/bin/env python3

# run_model_combinations.py
# Run a set of combined models using shared_ancestry_simulator.R based on a CSV file of model parameter combinations

# Written for "Evaluating statistics for the identification of introgressed loci"
# by Simon H. Martin, John W. Davey and Chris D. Jiggins
# Simon Martin: shm45@cam.ac.uk
# John Davey:   jd626@cam.ac.uk
# November-December 2013
# Revised February-May 2014, August 2014

import argparse
from math import floor
from datetime import datetime
from os import mkdir, chdir
from collections import defaultdict
from subprocess import call

parser = argparse.ArgumentParser(description='''Run a set of combined models using shared_ancestry_simulator.R.

    Takes a CSV file of model parameter combinations as input''')

parser.add_argument('-m', '--modelfile',  type=str, required=True)
parser.add_argument('-w', '--windows',  type=int, default=1000)
parser.add_argument('-t', '--threads', type=int, default=1)
parser.add_argument('-s', '--scalerate', type=float, default = 0.01)
parser.add_argument('-r', '--recombination', type=int, default = None)
parser.add_argument('-p', '--pathtosimulator', type=str, default="..")
parser.add_argument('-l', '--lengthofwindow', type=int, default = 5000)

args= parser.parse_args()

model_spec = """seqgen:
    m: HKY"""

model_spec += "\n    l: " + str(args.lengthofwindow)
model_spec += "\n    s: " + str(args.scalerate)
model_spec += """
ms:
    pops:
        - {name: 1, seqs: 8}
        - {name: 2, seqs: 8}
        - {name: 3, seqs: 8}
        - {name: 4, seqs: 8}
    opts:"""

if args.recombination != None:
    model_spec += "\n        recombination: " + str(args.recombination)

model_spec += """
        split:
            - {units: 3,   pops: [4,1]}
"""

def output_models(model_spec, model_type, params, log):
    pops = '0,0'
    if model_type is "Background":
        pops = '2,1'
    elif model_type is "Alternate":
        pops = '2,3'

    printpops = ''.join(pops.split(','))

    for model, model_windows in params.items():
        t123, t2x = model.split('/')

        model_t123 = "            - {units: " + t123 + ", pops: [3,1]}\n"
        model_txx =  "            - {units: " + t2x + ", pops: [" + pops + "]}\n"

        model_def = '{}_t123-{}_t{}-{}'.format(model_type, t123, printpops, t2x)
        try:
            model = open('{}.yml'.format(model_def), 'w')
            model.write(model_spec + model_t123 + model_txx)
            model.close()
        except:
            raise

        command = "{}/shared_ancestry_simulator.R -w {} -t {} -c {}.yml:1".format(
                    args.pathtosimulator, model_windows, args.threads, model_def)

        log.write('{}\t{}\n'.format(command,datetime.now()))
        try:
            call(command, shell=True)
        except:
            raise


def make_model_windows(args, log):
    try:
        models = open("../{}".format(args.modelfile))
    except:
        raise

    background = defaultdict(int)
    alternate = defaultdict(int)
    combinations = []

    header = models.readline().rstrip()
    
    for line in models:
        modeltype, bt123, bt21, at123, at23, altprop = line.rstrip().split(',')

        altprop = float(altprop)
        bg_windows = int(args.windows * (1-altprop))
        alt_windows = int(args.windows * altprop)
        combinations.append({'modeltype': modeltype,
                             'bt123': bt123, 'bt21': bt21,
                             'at123': at123, 'at23': at23,
                             'bg_windows': bg_windows,
                             'alt_windows':alt_windows})

        if bt123 != 'NA':
            pars = '{}/{}'.format(bt123,bt21)
            background[pars] = max(bg_windows, background[pars])
        if at123 != 'NA':
            pars = '{}/{}'.format(at123,at23)
            alternate[pars] = max(alt_windows, alternate[pars])

    output_models(model_spec, "Background", background, log)
    output_models(model_spec, "Alternate", alternate, log)

    return combinations, background, alternate


def write_windows(combf, model_filename, windows, startwin):
    modelf = open(model_filename)
    header = modelf.readline()
    if startwin == 0:
        combf.write(header)
    
    for win_i in range(0, windows):
        model = modelf.readline()
        block, blockNumber, blockName, *modelpars = model.split(',')
        blockNumber = int(blockNumber) + startwin
        
        outnum = '"{0}",{0},{0}'.format(blockNumber)
        
        combf.write(','.join([outnum] + modelpars))

    modelf.close()


def make_model_combinations(args, combinations, background, alternate):
    for comb in combinations:
        if comb['at123'] == 'NA': continue
            
        alt_pars = '{}/{}'.format(comb['at123'], comb['at23'])
        bg_pars = '{}/{}'.format(comb['bt123'], comb['bt21'])
        alt_spec = 'Alternate_t123-{}_t23-{}'.format(comb['at123'], comb['at23'])
        bg_spec = 'Background_t123-{}_t21-{}'.format(comb['bt123'], comb['bt21'])
            
        for seqs in 'ms', 'sg':
            combfilename = '{}.{}.{}.{}.csv'.format(alt_spec, bg_spec, args.windows, seqs)
            bg_filename = '{}.{}.{}.csv'.format(bg_spec, background[bg_pars], seqs)
            alt_filename = '{}.{}.{}.csv'.format(alt_spec, alternate[alt_pars], seqs)
            
            combf = open(combfilename,'w')
            write_windows(combf, bg_filename, comb['bg_windows'], 0)
            write_windows(combf, alt_filename, comb['alt_windows'], comb['bg_windows'])
            combf.close()

date = str(datetime.today()).split()[0]
outdir = "model_files_win{}_s{}_l{}_r{}_{}".format(args.windows, args.scalerate, args.lengthofwindow, args.recombination, date)

try:
    mkdir(outdir)
    chdir(outdir)
    log = open('{}.log'.format(outdir), 'w')
except:
    raise


combinations, background, alternate = make_model_windows(args, log)
make_model_combinations(args, combinations, background, alternate)